### PR TITLE
Add a dio_test project and move request integration test

### PR DESCRIPTION
**Pull request by kuhnroyal**
This was supposed to be just an attempt but while moving the integration test, I found a bug since one of the tests failed for the `Http2Adapter`.
So I think this is a good approach, I will move more tests in future PRs.

* run tests for all adapters
* fix redirect bug for `Http2Adapter`

### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dev/documentation/dio/latest/)
- [x] I have searched for a similar pull request in the [project](https://github.com/cfug/dio/pulls) and found none
- [x] I have updated this branch with the latest `main` branch to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I'm adding
- [x] I have updated the documentation (if necessary)
- [x] I have run the tests without failures
- [x] I have updated the `CHANGELOG.md` in the corresponding package

### CODE REVIEW

**Comment by kuhnroyal** [comment]
@AlexV525 What do you think about this approach? There is also a bugfix for redirects in here.

File: dio_tests/pubspec.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+name: dio_tests
+version: 1.0.0
+
+publish_to: none
+
+environment:
+  sdk: '>=3.0.0 <4.0.0'
```
[the remark this answers is not in the document]
Comment (AlexV525) [reply dio_tests/pubspec.yaml:7]: I mean we cannot run the `dio/test/request_integration_test.dart` anymore on 2.15 after this, correct?

**Comment by AlexV525** [comment]
> There is also a bugfix for redirects in here.

Can we extract the fix to another request first?

File: dio_tests/pubspec.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (kuhnroyal) [reply dio_tests/pubspec.yaml:7]: Yes, I don't see a way to run the same test suite on lower SDK if one of our adapters requires 3.0.0.
Unless we can maybe run the same tests with different `pubspec.yaml` files?

**Comment by kuhnroyal** [comment]
> > There is also a bugfix for redirects in here.
> 
> Can we extract the fix to another request first?

Done. See #2101

File: dio_tests/pubspec.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (kuhnroyal) [reply dio_tests/pubspec.yaml]: It actually works with a shared util package that we can import as dev dependency.
This way we can run tests on all supported SDK versions.

**Comment by AlexV525** [comment]
@kuhnroyal Still WIP?

File: dio_tests/pubspec.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (kuhnroyal) [reply dio_tests/pubspec.yaml]: Update this a bit but we now have a shared `dio_test` package that is imported as dev dependency into `dio` and `http2_adapter`. So the tests can be run with all Dart SDK targets.

**Comment by kuhnroyal** [comment]
I moved some tests in to the test package but I think we should move the rest in separate PRs. So I think this is good to go if we want to take that route.

**Review by AlexV525 — APPROVED** [review]
Thanks for all the work!

**Pull request merged**